src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Instance.hs

```diff
@@ -28,75 +38,85 @@ import           ZkFold.Symbolic.Data.Eq
-
-instance Arithmetic a => AdditiveMonoid (ArithmeticCircuit a) where
-    zero = circuit $ newAssigned (const zero)
-
-instance Arithmetic a => AdditiveGroup (ArithmeticCircuit a) where
-    negate r = circuit $ do
-        i <- runCircuit r
-        newAssigned (\x -> negate (x i))
-
-    r1 - r2 = circuit $ do
-        i <- runCircuit r1
-        j <- runCircuit r2
-        newAssigned (\x -> x i - x j)
-
-instance Arithmetic a => MultiplicativeSemigroup (ArithmeticCircuit a) where
-    r1 * r2 = circuit $ do
-        i <- runCircuit r1
-        j <- runCircuit r2
-        newAssigned (\x -> x i * x j)
-
-instance Arithmetic a => Exponent (ArithmeticCircuit a) Natural where
+instance (Arithmetic a, KnownNat n) => SymbolicData a n (ArithmeticCircuit n a) where
+    pieces = id
+
+    restore = ArithmeticCircuit
+
+-- TODO: I had to add these constraints and I don't like them
+instance
+    ( KnownNat (n * Order a)
+    , KnownNat (Log2 ((n * Order a) - 1) + 1)
+    ) => Finite (ArithmeticCircuit n a) where
+    type Order (ArithmeticCircuit n a) = n * Order a
+
+instance (Arithmetic a, KnownNat n) => AdditiveSemigroup (ArithmeticCircuit n a) where
+    r1 + r2 = circuitN $ do
+        is <- runCircuit r1
+        js <- runCircuit r2
+        for (Z.zip is js) $ \(i, j) -> newAssigned (\x -> x i + x j)
+
+instance (Arithmetic a, Scale c a, KnownNat n) => Scale c (ArithmeticCircuit n a) where
+    scale c r = circuitN $ do
+        is <- runCircuit r
+        for is $ \i -> newAssigned (\x -> (c `scale` one :: a) `scale` x i)
+
+instance (Arithmetic a, KnownNat n) => AdditiveMonoid (ArithmeticCircuit n a) where
+    zero = circuitN $ Vector <$> replicateM (Haskell.fromIntegral $ value @n) (newAssigned (const zero))
+
+instance (Arithmetic a, KnownNat n) => AdditiveGroup (ArithmeticCircuit n a) where
+    negate r = circuitN $ do
+        is <- runCircuit r
+        for is $ \i -> newAssigned (\x -> negate (x i))
+
+    r1 - r2 = circuitN $ do
+        is <- runCircuit r1
+        js <- runCircuit r2
+        for (Z.zip is js) $ \(i, j) -> newAssigned (\x -> x i - x j)
+
+instance (Arithmetic a, KnownNat n) => MultiplicativeSemigroup (ArithmeticCircuit n a) where
+    r1 * r2 = circuitN $ do
+        is <- runCircuit r1
+        js <- runCircuit r2
+        for (Z.zip is js) $ \(i, j) -> newAssigned (\x -> x i * x j)
+
+instance (Arithmetic a, KnownNat n) => Exponent (ArithmeticCircuit n a) Natural where
     (^) = natPow
 
-instance Arithmetic a => MultiplicativeMonoid (ArithmeticCircuit a) where
+instance (Arithmetic a, KnownNat n) => MultiplicativeMonoid (ArithmeticCircuit n a) where
     one = mempty
 
-instance (Arithmetic a, FromConstant b a) => FromConstant b (ArithmeticCircuit a) where
-    fromConstant c = embed (fromConstant c)
+-- TODO: The constant will be replicated in all outputs. Is this the desired behaviour?
+instance (Arithmetic a, FromConstant b a, KnownNat n) => FromConstant b (ArithmeticCircuit n a) where
```

Comment:
Do we want this instance? Or just the one for `ArithmeticCircuit 1 a`?

Reply:
We need this as `Semiring x` requires `FromConstant Natural x` and `Ring x` requires `FromConstant Integer x`

Reply:
BTW, I guess we can introduce the following instances:
```haskell
instance (FromConstant b a, KnownNat n) => FromConstant b (Vector n a) -- by replicate
instance (FromConstant b (Vector n a)) => FromConstant b (ArithmeticCircuit n a) -- by embed
```

Reply:
Yes, on second thought, it all makes sense and might be useful. For example, if we need to add the same constant to a bunch of variables at once.

Reply:
Having the algebraic instances for n/=1 feels weird to me